theseus/tests/test_theseus_layer.py

```diff
@@ -280,11 +301,44 @@ def cost_weight_fn():
             | (info.status == th.NonlinearOptimizerStatus.FAIL)
         ).all()
 
-        loss = F.mse_loss(pred_vars["coefficients"], target_vars["coefficients"])
+        if learning_method == "leo":
+            # groundtruth cost
+            x_gt = target_vars["coefficients"]
+            input_values_gt = {
+                "coefficients": x_gt,
+                cost_weight_param_name: cost_weight_fn(),
+            }
+            layer_to_learn.objective.update(input_values_gt)
+            cost_gt = torch.sum(layer_to_learn.objective.error(), dim=1)  # B x 1
+            cost_gt = cost_gt.to(device)
+
+            # optimizer cost
+            x_opt = pred_vars["coefficients"].detach()
+            x_samples = layer_to_learn.optimizer.compute_samples(
+                n_samples=10, T=1.0
+            )  # B x N x S
+            if x_samples is None:  # use mean solution
+                x_samples = x_opt.reshape(x_opt.shape[0], -1).unsqueeze(-1)  # B x N x 1
+            cost_opt = get_average_sample_cost(
+                x_samples, layer_to_learn, cost_weight_param_name, cost_weight_fn
+            )
+            cost_opt = cost_opt.to(device)
+
+            # loss value
+            l2_reg = F.mse_loss(
+                cost_weight_fn(), torch.zeros((1, num_points), device=device)
+            )
+            loss = (cost_gt - cost_opt) + 10.0 * l2_reg
+            loss = torch.mean(loss, dim=0)
+        else:
+            loss = F.mse_loss(pred_vars["coefficients"], target_vars["coefficients"])
+
         loss.backward()
-        print(i, loss.item(), loss.item() / loss0)
         optimizer.step()
-        if loss.item() / loss0 < 5e-3:
+
+        mse_loss = F.mse_loss(pred_vars["coefficients"], target_vars["coefficients"])
```

Comment:
Instead of calculating the mse loss again you can shuffle things around.

```
mse_loss = F.mse_loss(pred_vars["coefficients"], target_vars["coefficients"])
....
if learning_method == "leo":
...
else:
    loss = mse_loss

loss.backward()
optimizer.step()

print(i, mse_loss.item())        
if mse_loss.item() / loss0 < 5e-3:
    solved = True
    break
...
```

Reply:
Agreed, moved for both test and state_estimation example.